url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+			    rLinha = row;
+			    minDegree = degree;
+			    isHDPC2 = isHDPC;
+			} else
+			    continue;
+		    } else {
+			r = nonZeros;
+			rLinha = row;
+			minDegree = degree;
+			isHDPC2 = isHDPC;
+		    }
+		}
+	    }
+	    
+	    if ( r == L + 1 ) // DECODING FAILURE
+		throw new SingularMatrixException("Decoding Failure - PI Decoding @ Phase 1: All entries in V are zero.");
+	    
+	    /*
+	     * choose the row
+	     */
+	    
+	    if ( r != 2 ) {
+		// "HDPC rows should not be chosen untill all non-HDPC rows have been processed"
+		if ( isHDPC2 && chosenRowsCounter < nonHDPCRows ) { // if it is, then we must choose another line
+		
+		    int newDegree = 256 * L;
+		    int newR = L + 1;
+		    
+		    // lets go search all the rows
+		    for (Row row : rows.values()) {
+			// if they're not HDPC rows
+			if ( !row.isHDPC && row.degree != 0 ) {
+			    // does it have less non-zeros than our 'currently chosen' one?
+			    if ( row.nonZeros < newR ) {// if it does, since it's non-HDPC it'll also have a lower degree
+			    
+				// so lets update our 'currently chosen' row
+				newR = row.nonZeros;
+				rLinha = row.id;
+				newDegree = row.degree;
+			    }
+			} else
+			    continue;
+		    }
+		}
+		
+		// choose rLinha
+		chosenRowsCounter++;
+	    } else { // r = 2
+	    
+		// do we have a row with exactly two 1's? (remember we already know that there are rows with only 2 non-zeros)
+		if ( minDegree == 2 ) {
+		    /*
+		     * create graph
+		     */
+		    
+		    // allocate memory
+		    Map<Integer, Set<Integer>> graph = new HashMap<Integer, Set<Integer>>();
+		    
+		    // lets go through all the rows... (yet again!)
+		    for (Row row : rows.values()) {
+			// is this row an edge?
+			if ( row.edges != null ) {
+			    // get the nodes connected through this edge
+			    Integer[] edge = row.edges.toArray(new Integer[2]);
+			    int node1 = edge[0];
+			    int node2 = edge[1];
+			    
+			    // node1 already in graph?
+			    if ( graph.keySet().contains(node1) ) { // it is
+			    
+				// then lets add node 2 to its neighbours
+				graph.get(node1).add(node2);
+			    } else { // it isn't
+			    
+				// allocate memory for its neighbours
+				Set<Integer> edges = new HashSet<Integer>();
+				
+				// add node 2 to its neighbours
+				edges.add(node2);
+				
+				// finally, add node 1 to the graph along with its neighbours
+				graph.put(node1, edges);
+			    }
+			    
+			    // node2 already in graph?
+			    if ( graph.keySet().contains(node2) ) { // it is
+			    
+				// then lets add node 1 to its neighbours
+				graph.get(node2).add(node1);
+			    } else { // it isn't
+			    
+				// allocate memory for its neighbours
+				Set<Integer> edges = new HashSet<Integer>();
+				
+				// add node 1 to its neighbours
+				edges.add(node1);
+				
+				// finally, add node 2 to the graph along with its neighbours
+				graph.put(node2, edges);
+			    }
+			} else
+			    continue;
+		    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1343:1386, 2258:2290.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.